File: test-suite/src/test/java/io/micronaut/crac/OrderedResourceCheckpointSimulatorTest.java
package io.micronaut.crac;

import io.micronaut.context.BeanContext;
import io.micronaut.context.annotation.Property;
import io.micronaut.crac.test.CheckpointSimulator;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

@MicronautTest
@Property(name = "spec.name", value = OrderedResourceCheckpointSimulatorTest.SPEC_NAME)
public class OrderedResourceCheckpointSimulatorTest {

    public static final String SPEC_NAME = "OrderedResourceCheckpointSimulatorTest";

    //tag::test[]
    @Inject
    BeanContext ctx;

    @Test
    void testCustomOrderedResourceUsingCheckpointSimulator() {
        ResourceBean myBean = ctx.getBean(ResourceBean.class);
        CheckpointSimulator checkpointSimulator = ctx.getBean(CheckpointSimulator.class); // <1>
        assertTrue(myBean.isRunning());

        checkpointSimulator.runBeforeCheckpoint();  // <2>
        assertFalse(myBean.isRunning());

        checkpointSimulator.runAfterRestore(); // <3>
        assertTrue(myBean.isRunning());
    }
    //end::test[]
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/crac/OrderedResourceCheckpointSimulatorTest.kt
package io.micronaut.crac

import io.micronaut.context.BeanContext
import io.micronaut.context.annotation.Property
import io.micronaut.crac.test.CheckpointSimulator
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import jakarta.inject.Inject
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

@MicronautTest
@Property(name = "spec.name", value = OrderedResourceCheckpointSimulatorTest.SPEC_NAME)
class OrderedResourceCheckpointSimulatorTest {
    companion object {
        const val SPEC_NAME = "OrderedResourceCheckpointSimulatorTest"
    }

    //tag::test[]
    @field:Inject
    lateinit var ctx: BeanContext

    @Test
    fun testCustomOrderedResourceUsingCheckpointSimulator() {
        val myBean = ctx.getBean(ResourceBean::class.java)
        val checkpointSimulator = ctx.getBean(CheckpointSimulator::class.java) // <1>
        Assertions.assertTrue(myBean.isRunning)

        checkpointSimulator.runBeforeCheckpoint() // <2>
        Assertions.assertFalse(myBean.isRunning)

        checkpointSimulator.runAfterRestore() // <3>
        Assertions.assertTrue(myBean.isRunning)
    }
    //end::test[]
}
